libraries/AP_Radio/AP_Radio_cypress.cpp

```diff
@@ -0,0 +1,1748 @@
+
+    // If COMPLETE and ERROR bits mismatch, then re-read register
+    if ((ret & (CYRF_RXC_IRQ | CYRF_RXE_IRQ)) != 0
+        && (ret & (CYRF_RXC_IRQ | CYRF_RXE_IRQ)) != (CYRF_RXC_IRQ | CYRF_RXE_IRQ)) {
+        uint8_t v2;
+        dev->read(&v2, 1);
+        ret |= v2;   // re-read and make bits sticky
+    }
+    dev->set_chip_select(false);
+    return ret;
+}
+
+/*
+  force the initial state of the radio
+ */
+void AP_Radio_cypress::force_initial_state(void)
+{
+    while (true) {
+        write_register(CYRF_XACT_CFG, CYRF_FRC_END);
+        uint32_t start_ms = AP_HAL::millis();
+        do {
+            if ((read_register(CYRF_XACT_CFG) & CYRF_FRC_END) == 0) {
+                return;                     // FORCE_END done (osc running)
+            }
+        } while (AP_HAL::millis() - start_ms < 5);
+
+        // FORCE_END failed to complete, implying going SLEEP to IDLE and
+        // oscillator failed to start.  Recover by returning to SLEEP and
+        //  trying to start oscillator again.
+        write_register(CYRF_XACT_CFG, CYRF_MODE_SLEEP);
+    }
+}
+
+/*
+  set desired channel
+ */
+void AP_Radio_cypress::set_channel(uint8_t channel)
+{
+    if (dsm.forced_channel != -1) {
+        channel = dsm.forced_channel;
+    }
+    write_register(CYRF_CHANNEL, channel);
+}
+
+void AP_Radio_cypress::radio_set_config(const struct config *conf, uint8_t size)
+{
+    // setup required radio config
+    for (uint8_t i=0; i<size; i++) {
+        write_register(conf[i].reg, conf[i].value);
+    }
+}
+
+/*
+  initialise the radio
+ */
+void AP_Radio_cypress::radio_init(void)
+{
+    Debug(1, "Cypress: radio_init starting\n");
+
+    // wait for radio to settle
+    uint16_t i;
+    for (i=0; i<1000; i++) {
+        uint8_t chan = read_register(CYRF_CHANNEL);
+        if (chan == 1) {
+            break;
+        }
+        write_register(CYRF_CHANNEL, 1);
+        hal.scheduler->delay(10);
+    }
+    if (i == 1000) {
+        Debug(1, "Cypress: radio_init failed\n");
+        return;
+    }
+
+    // base config
+    radio_set_config(cyrf_config, ARRAY_SIZE(cyrf_config));
+
+    // start with receive config
+    radio_set_config(cyrf_transfer_config, ARRAY_SIZE(cyrf_transfer_config));
+
+    if (get_disable_crc()) {
+        write_register(CYRF_RX_OVERRIDE, CYRF_DIS_RXCRC);
+    }
+    
+    dsm_setup_transfer_dsmx();
+
+    write_register(CYRF_XTAL_CTRL,0x80);  // XOUT=BitSerial
+    force_initial_state();
+    write_register(CYRF_PWR_CTRL,0x20);   // Disable PMU
+
+    // start in RECV state
+    state = STATE_RECV;
+
+    Debug(1, "Cypress: radio_init done\n");
+
+    start_receive();
+
+    // setup handler for rising edge of IRQ pin
+#if CONFIG_HAL_BOARD == HAL_BOARD_PX4
+    stm32_gpiosetevent(CYRF_IRQ_INPUT, true, false, false, irq_radio_trampoline);
+#elif CONFIG_HAL_BOARD == HAL_BOARD_CHIBIOS
+    hal.gpio->attach_interrupt(HAL_GPIO_RADIO_IRQ, trigger_irq_radio_event, HAL_GPIO_INTERRUPT_RISING);
+#endif
```

Comment:
this is another one of those cases where a hal should mean this block of #if code wouldn't be necerssary. 
Can't hal->attach_interrupt(....) be made smart enough to know if it's on px4, and then cal stm32_gpiosetevent behin the scenes?